File: go/hello.go
package main

import (
    "fmt"
    "math/big"
    "time"
)

var multiplicationCount int // Variable to count the number of multiplications

// Factorial function that calculates the factorial recursively
func factorial(n int) *big.Int {
    if n == 0 {
        return big.NewInt(1)
    } else {
        multiplicationCount++ // Increment the count for each multiplication
        return new(big.Int).Mul(big.NewInt(int64(n)), factorial(n-1))
    }
}

// Function to count the number of digits in a big.Int
func countDigits(num *big.Int) int {
    return len(num.String()) // Convert to string and get its length
}

func main() {
    startTime := time.Now()
    result := factorial(10000)
    endTime := time.Now()

    digitCount := countDigits(result) // Get the number of digits in the result

    fmt.Printf("Factorial has %d digits.\n", digitCount) // Print the number of digits
    fmt.Printf("Number of multiplications: %d\n", multiplicationCount) // Log the number of multiplications
    fmt.Printf("Time taken: %d milliseconds\n", endTime.Sub(startTime).Milliseconds()) // Print in milliseconds
}

File: js/hello.js
let multiplicationCount = 0;  // Initialize the multiplication counter

function factorial(n) {
    if (n === 0) {
        return BigInt(1);
    } else {
        multiplicationCount++;  // Increment the counter for each multiplication
        return BigInt(n) * factorial(n - 1);
    }
}

// Function to calculate the number of digits in a BigInt
function countDigits(bigInt) {
    return bigInt.toString().length; // Convert to string and get its length
}

const startTime = new Date().getTime();
const result = factorial(10000);  // Calculate factorial using BigInt
const endTime = new Date().getTime();

const digitCount = countDigits(result); // Get the number of digits in the result

console.log(`Factorial has ${digitCount} digits.`);
console.log(`Number of multiplications: ${multiplicationCount}`);  // Log the number of multiplications
console.log(`Time taken: ${endTime - startTime} milliseconds`); // Log the time taken
